### PR TITLE
Initial CI migration to Github Actions

### PR DESCRIPTION
**Pull request by WillB97**
An attempt to directly convert the CircleCI config to github actions

Current issues
- The build artifacts are additonally zipped
- webots docker image takes several minutes to initialise, on every run

We could partially overcome the first issue by making use of actions/upload-release-asset to directly place the artifact in the release.
~The second issue may be reduced by making a docker image on the github container repository.~

### CODE REVIEW

**Comment by WillB97** [comment]
Moving to running Webots directly instead of in docker seems to greatly improve the spin up time to around 80s, all caching attempts I've tried so far have had little benefit on cache-hit and major detriments on misses.

With the build artifacts, moving to automatically deplying them to releases seems like a good solution. With people being able to remove the outer archive if they want the build of a non-tag. A possible solution of just uploading the unarchived build turns out to have shortcomings in that execution permissions are dropped and each file is uploaded as a separate HTTP requests leading to notable overhead.

**Comment by PeterJCLaw** [comment]
Do you have an example of what you mean by the build uploads?
We have one file to upload -- the (custom built) archive of the repo (which I wouldn't expect to preserve execution permissions anyway), so I'm not sure I understand the issues you're hitting?

**Comment by WillB97** [comment]
With github actions artifact even uploading a single zip gets encapsulated within another zip. While you could extract the repo archive and rely on the artifact handling to zip it, you loose the execute permission on all the scripts (which probably isn't a big issue) and the artifact's zipping is significantly slower.

Since the system for uploading files to a release uses a different backend, it doesn't add the encapsulating zip. But I haven't figured out all the syntax for it yet.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        venv/bin/pip install -r all-requirements.txt
+    - name: Lint with flake8
+      run: |
+        source venv/bin/activate
+
+        # Ignore TODOs on CI, for now, even though we do want them
+        # highlighted in development.
+        ./script/linting/lint --extend-ignore=T000
+
+  typecheck:
```
Comment (RealOrangeOne) [inline .github/workflows/tests.yml:94]: Given the steps between linting and typechecking are nearly identical, probably makes sense to just combine the steps?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,153 @@
+name: Simulator tests
+
+on:
+  push: {}
+  release:
+    types: [ created ]
```
Comment (RealOrangeOne) [inline .github/workflows/tests.yml:6]: Why are we running this when creating releases too?

If the intention is to make publishing better, I'd suggest creating a dedicated workflow for it which only runs on release "publish", and uses https://github.com/RealOrangeOne/date-group/blob/master/.github/workflows/release.yml#L44

This could then be done instead of the archiving stuff below which creates double zips, or I guess as well as if we want to be sure it's working correctly.

File: .github/workflows/run-match.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+          export DEBIAN_FRONTEND=noninteractive
+          sudo apt-get update
+          sudo apt-get install --yes wget zip pulseaudio
+      - name: Download Webots
+        run: |
+          wget -O ./webots.deb \
+          https://github.com/cyberbotics/webots/releases/download/R${{ matrix.webots-version }}/webots_${{ matrix.webots-version }}_amd64.deb
+      - name: Install Webots & Xvfb
+        run: |
+          sudo apt-get install --yes xvfb ./webots.deb
```
Comment (RealOrangeOne) [inline .github/workflows/run-match.yml:35]: Probably worth just installing `xvfb` with the installs on line 28? Keep things separate?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (WillB97) [reply .github/workflows/tests.yml:94]: We could also merge test and essentially run the check script against all python versions

File: .github/workflows/run-match.yml (hunk 3 above)
Comment (WillB97) [reply .github/workflows/run-match.yml:35]: Certainly, this was just a hold over from some caching attempts

File: .github/workflows/tests.yml (hunk 2 above)
Comment (WillB97) [reply .github/workflows/tests.yml]: Is it worth also disabling run-match on releases?

**Comment by WillB97** [comment]
We should probably make a draft release on this branch to test that CI works before merging

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    - uses: actions/checkout@v2
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.7
+    - name: Build archive
+      id: build_archive
+      run: |
+        mkdir archive
+        NAME="competition-simulator-$(git describe --always --tags).zip"
```
Comment (RealOrangeOne) [inline .github/workflows/publish.yml:20]: You can probably get the tag from `github.ref` much easier, and then perhaps use it instead of the `set-output` stuff below

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
+name: Publish simulator archive
+
+on:
+  release:
+    types: [ created ]
```
Comment (RealOrangeOne) [inline .github/workflows/publish.yml:5]: You'd be better off using published here, as that means that drafting releases doesn't trigger this, and then it'll be triggered on the final publish. Creating without drafting implies publishing.

```suggestion
    types: [ published ]
```

src: https://github.com/RealOrangeOne/zoloto/blob/de61a0a2068f51a316ba25a6a435336029b6706d/.github/workflows/pypi.yml#L5

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    - name: Upload binaries to release
+      uses: svenstaro/upload-release-action@v1-release
+      with:
+        repo_token: ${{ secrets.GITHUB_TOKEN }}
+        file: archive/${{ steps.build_archive.outputs.archive }}
+        asset_name: ${{ steps.build_archive.outputs.archive }}
+        tag: ${{ github.ref }}
+        overwrite: true
+
+  build:
```
Comment (RealOrangeOne) [inline .github/workflows/publish.yml:32]: Is there a reason to build _and_ publish here? Once it's in the issue it's probably fine.

File: .github/workflows/run-match.yml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  release:
+    types: [ created ]
+
+jobs:
+  run-comp-match:
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: bash
+        working-directory: ${{ github.workspace }}
```
Comment (RealOrangeOne) [inline .github/workflows/run-match.yml:14]: Is this stuff necessary?

File: .github/workflows/run-match.yml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+          export DEBIAN_FRONTEND=noninteractive
+          sudo apt-get update
+          sudo apt-get install --yes wget zip pulseaudio xvfb
+      - name: Download Webots
+        run: |
+          wget -O ./webots.deb \
+          https://github.com/cyberbotics/webots/releases/download/R${{ matrix.webots-version }}/webots_${{ matrix.webots-version }}_amd64.deb
+      - name: Install Webots
+        run: |
+          sudo apt-get install --yes ./webots.deb
```
Comment (RealOrangeOne) [inline .github/workflows/run-match.yml:35]: Skip the `apt-get` wrapper?

```suggestion
          sudo dpkg -i ./webots.deb
```

File: .github/workflows/run-match.yml [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+      - name: Prepare team code archive
+        run: |
+          mkdir archives
+          cd archives
+
+          # Convert the sleep comments to print statement so there's a way to
+          # gauge progress and so that Circle doesn't shut things down due to
```
Comment (RealOrangeOne) [inline .github/workflows/run-match.yml:42]:
```suggestion
          # gauge progress and so that GitHub doesn't shut things down due to
```

File: README.md [hunk 10]
```diff
@@ -125,8 +126,7 @@ zip ABC.zip robot.py
 ## Doing a release
 
 1. Create a new tag & push
-2. Wait for CI to build a zip archive
-3. Upload the archive to the GitHub release for the tag
+2. Wait for CI to build a zip archive and upload it to the tag
```
Comment (RealOrangeOne) [inline README.md:129]: Not quite true. On top of pushing a tag, you also need to create a release in GitHub

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -0,0 +1,69 @@
+name: Simulator tests
+
+on:
+  push: {}
+
+jobs:
+  run-checks:
+    runs-on: ubuntu-latest
```
Comment (RealOrangeOne) [inline .github/workflows/tests.yml:8]: We should run on all OS'

File: .github/workflows/run-match.yml (hunk 7 above)
Comment (WillB97) [reply .github/workflows/run-match.yml:14]: Only when using docker otherwise `/bin/sh` is used

File: .github/workflows/run-match.yml (hunk 8 above)
Comment (PeterJCLaw) [reply .github/workflows/run-match.yml:35]: Note that `dpkg` won't install dependencies, `apt-get` will do.

File: .github/workflows/publish.yml (hunk 4 above)
Comment (WillB97) [reply .github/workflows/publish.yml:20]: We need the tag basename and `github.ref` gives the full ref name

File: .github/workflows/run-match.yml (hunk 8 above)
Comment (WillB97) [reply .github/workflows/run-match.yml]: This was indeed the case

File: modules/controller_utils/tests.py [hunk 12]
```diff
@@ -30,7 +30,7 @@ def fake_tla() -> str:
 
 @contextlib.contextmanager
 def mock_match_file() -> Iterator[IO[str]]:
-    with tempfile.NamedTemporaryFile(suffix='.json', mode='r+t') as f:
+    with tempfile.NamedTemporaryFile(suffix='.json', mode='r+t', delete=False) as f:
```
Comment (PeterJCLaw) [inline modules/controller_utils/tests.py:33]: This will leave the files around and clutter up developers machines. The point of using the context manager with a temp file is to prevent that. Why is this needed?

Comment (WillB97) [reply modules/controller_utils/tests.py:33]: Windows cannot handle `NamedTemporaryFile` being closed and reopened without it. The files are fractions of KB and will still be removed on reboot.

File: modules/sr/robot/coordinates/tests.py [hunk 13]
```diff
@@ -368,7 +368,9 @@ def test_polar(self) -> None:
 
         for cartesian, expected in cases:
             with self.subTest(cartesian):
-                self.assertEqual(expected, polar_from_cartesian(cartesian))
+                self.assertAlmostEqual(expected.length, polar_from_cartesian(cartesian).length)
+                self.assertAlmostEqual(expected.rot_x, polar_from_cartesian(cartesian).rot_x)
+                self.assertAlmostEqual(expected.rot_y, polar_from_cartesian(cartesian).rot_y)
```
Comment (PeterJCLaw) [inline modules/sr/robot/coordinates/tests.py:373]: If we really need to do this please could we avoid the duplicate calls to `polar_from_cartesian`?
However unfortunately this is a semantic change versus the original assertion and one which is not very future-proof.
Suggest that instead of this we perhaps force the rounding ourselves?
```suggestion
                actual = polar_from_cartesian(cartesian)
                # Cope with floating point differences
                rounded = PolarCoord(*(round(x, 7) for x in actual))

                self.assertEqual(expected, actual)
```

File: modules/controller_utils/tests.py (hunk 12 above)
Comment (PeterJCLaw) [reply modules/controller_utils/tests.py]: bcb1959f8953f37714da61eaab03775ab9ab625d

File: modules/sr/robot/coordinates/tests.py (hunk 13 above)
Comment (WillB97) [reply modules/sr/robot/coordinates/tests.py]: 87bd35c2aa9c3c0e502865d016b27f1251604d0c

File: .github/workflows/tests.yml [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+      if: steps.cache-dep.outputs.cache-hit != 'true' && matrix.test-os == 'windows-latest'
+      run: |
+        py -m venv venv
+        source venv/Scripts/activate
+        py -m pip install -U pip
+        pip install -r all-requirements.txt
+    - name: Define scripts to test
+      id: define_scripts
+      run: |
+        SCRIPTS=$(find script -type f | grep -vE "/(linting/|testing/|typing/|check$)"|tr "\n" " ")
```
Comment (PeterJCLaw) [inline .github/workflows/tests.yml:58]:
```suggestion
        SCRIPTS=$(find script -type f | grep -vE "/(linting/|testing/|typing/|check$)" | tr "\n" " ")
```

**Review by PeterJCLaw — APPROVED** [review]
